### PR TITLE
Login dialog and auth system

### PR DESCRIPTION
**Pull request by arlowhite**
* Add login dialog and profile button.
* auth service and HTTP interceptor that adds JWT Authorization token.
* Hide UI when not logged in.
* ESRI request interceptor, add token to reef guide tile URLs

image

image

### CODE REVIEW

File: src/app/app.config.ts [hunk 1]
```diff
@@ -1,15 +1,19 @@
...
 
 export const appConfig: ApplicationConfig = {
   providers: [
     provideExperimentalZonelessChangeDetection(),
     provideRouter(routes, withComponentInputBinding()),
     provideAnimationsAsync(),
-    provideHttpClient(withFetch())
+    provideHttpClient(
+      withFetch(),
+      withInterceptors([authInterceptor])
```
Comment (arlowhite) [inline src/app/app.config.ts:16]: This is adding the auth interceptor to the default `HttpClient`. Currently, this is only injected into the API services, but in the future if we need an `HttpClient` for other purposes, I'll make the default unauthenticated, and setup another just for use in the APIs.

File: src/app/location-selection/reef-guide-map.service.ts [hunk 2]
```diff
@@ -67,7 +71,43 @@ export class ReefGuideMapService {
...
+  private setupRequestInterceptor() {
+    const apiUrl = environment.reefGuideApiUrl;
+    esriConfig.request.interceptors!.push({
+      urls: [
+        new RegExp(`^${apiUrl}`)
+      ],
+      before: params => {
+        const url = params.url as string;
+        // security double-check that we're only intercepting our API urls.
+        if (url.startsWith(apiUrl)) {
```
Comment (arlowhite) [inline src/app/location-selection/reef-guide-map.service.ts:87]: this should prevent accidentally adding token to some other ArcGIS request if their `url` filter doesn't work as expected.

File: src/app/auth/login-dialog/login-dialog.component.html
Comment (PeterBaker0) [inline src/app/auth/login-dialog/login-dialog.component.html:1]: Would be nice if pressing enter ran the login button action

Comment (PeterBaker0) [reply src/app/auth/login-dialog/login-dialog.component.html:1]: Not that critical tho

Comment (PeterBaker0) [inline src/app/auth/login-dialog/login-dialog.component.html:1]: Other thing I notice is that the state of the "register" / "login" button seems to persist when swapping between register and login. E.g. if you put in an invalid login, then swap to register, you can't click register (button is grayed out) until you change the text in the input fields.

File: src/api/api-util.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+      "name": "ZodError"
+    }
+  }
+]
+*/
+
+/**
+ * Get user-friendly error message from API error.
+ * @param error
+ */
```
Comment (PeterBaker0) [inline src/api/api-util.ts:28]: Okay - just noting this is just for the 400 error response from the zod middleware on express. 

The general error class is here

```
const errorResponse: ErrorResponse = {
    status: statusCode,
    message: message,
  };
```

I would prefer to have the validation error from zod report a 422 (parse error) - this is a bit easier to isolate. Anyway thanks for pulling this out.

File: src/app/app.config.ts (hunk 1 above)
Comment (PeterBaker0) [reply src/app/app.config.ts:16]: Cool

File: src/app/auth/login-dialog/login-dialog.component.html [hunk 4]
```diff
@@ -0,0 +1,32 @@
+@let m = mode();
+@let actionLabel = m === 'register' ? 'Register' : 'Login';
+<h2 mat-dialog-title>{{ actionLabel }}</h2>
+<mat-dialog-content>
+  @if (m === 'register') {
+    <p>Sign up with a new account.</p>
+  }
+<!-- FIXME enter on inputs (ngSubmit) not working -->
```
Comment (PeterBaker0) [inline src/app/auth/login-dialog/login-dialog.component.html:8]: oh see you already caught this

File: src/app/location-selection/reef-guide-map.service.ts [hunk 5]
```diff
@@ -67,7 +71,43 @@ export class ReefGuideMapService {
...
+              };
+            } else {
+              params.requestOptions.headers.Authorization = headerVal;
+            }
+          } else {
+            // All reef-guide-api layers require authentication, so spare the API
+            // from requests it would reject with a 401 anyway.
+            throw new Error("Not authenticated");
+            // Note: we also have access to an AbortSignal here.
+            // const signal = params.requestOptions.signal as AbortSignal;
```
Comment (PeterBaker0) [inline src/app/location-selection/reef-guide-map.service.ts:103]: makes sense

**Review by PeterBaker0 — APPROVED** [review]
Other than couple of bugs, looks good.

File: src/app/auth/login-dialog/login-dialog.component.html (hunk 4 above)
Comment (arlowhite) [reply src/app/auth/login-dialog/login-dialog.component.html:8]: ya, I'll try another way in next PR, it should just work with `(ngSubmit)`, but I think I could grab the event off the inputs directly.

**Pull request merged**